### PR TITLE
#3753 Inventory: Prepare Inventorylines shall update existing lines 

### PR DESCRIPTION
**Pull request by cristinamghita**
#3753 Inventory: Prepare Inventorylines shall update existing lines

### CODE REVIEW

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java [hunk 1]
```diff
@@ -93,6 +101,16 @@ private I_M_Inventory getInventory()
 		return _inventory;
 	}
 
+	private List<Integer> fetchExistentHUIds()
+	{
+		final I_M_Inventory inventory = getInventory();
+		return Services.get(IInventoryDAO.class).retrieveLinesForInventoryId(inventory.getM_Inventory_ID(), I_M_InventoryLine.class)
+				.stream()
+				.map(I_M_InventoryLine::getM_HU_ID)
+				.collect(ImmutableList.toImmutableList());
```
Comment (teosarca) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java]: before collect, pls `filter(huId -> huId > 0)`

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java [hunk 2]
```diff
@@ -117,35 +135,55 @@ private I_M_Inventory getInventory()
...
-		inventoryLine.setC_UOM(huProductStorage.getC_UOM());
 		inventoryLine.setQtyBook(huProductStorage.getQty());
 		inventoryLine.setQtyCount(huProductStorage.getQty());
 
 		InterfaceWrapperHelper.save(inventoryLine);
 
 		return inventoryLine;
 	}
+
+	private I_M_InventoryLine fetchInventoryLineForHU(final I_M_HU hu)
```
Comment (teosarca) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java]: generally, pls pls pls don't use models when simply the M_HU_ID would be enough!

**Comment by teosarca** [comment]
Please read what i've annotated on lines, but overall, i think the approach is not OK.
* you are loading all inventory lines just to fetch the M_HU_ID from them
* then while u are processing u are querying it again to find which is the inventory line which is the inventory line which has your HU. See `fetchInventoryLineForHU`. And you are doing this in a process while we agreed to have NO database calls from anything else than a DAO/Repository.

Suggestion to fix:
* load all inventory lines and index them by M_HU_ID. I think you can skip those which does not have a M_HU_ID set. You can memorize that map as a field in your process.
* While importing you just have to consult that map if you already have an inventory line for your HU

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java [hunk 3]
```diff
@@ -117,30 +129,40 @@ private I_M_Inventory getInventory()
...
 	{
 		return Services.get(IHandlingUnitsBL.class)
 				.getStorageFactory()
 				.streamHUProductStorages(hu)
-				.map(huProductStorage -> createInventoryLine(huProductStorage));
+				.map(huProductStorage -> createUpdateInventoryLine(huProductStorage, updateLine));
 	}
 
-	private I_M_InventoryLine createInventoryLine(final IHUProductStorage huProductStorage)
+	private I_M_InventoryLine createUpdateInventoryLine(@NonNull final IHUProductStorage huProductStorage, final boolean updateLine)
```
Comment (teosarca) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java]: the `updateLine` param is not needed. You could check inventoryLinesByHU directly and decide if create or update.

Comment (cristinamghita) [reply de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/process/M_Inventory_CreateCountLinesFromHU.java]: yep; right! thx!

**Pull request merged**